docs/config.rst
Configuration
=============

*Mesh Info* is configured via an environment file named ``.env`` in the same folder as the project was cloned into.
If you followed the :doc:`installation instructions <installation>` then it should be ``/opt/mesh-info/src/.env``.

Example
-------

.. note::

   The Tile URL below requires registering with Stadia Maps: https://stadiamaps.com/stamen/onboarding/create-account

.. code-block:: ini

   MESH_INFO_LOCAL_NODE="10.1.1.1"
   MESH_INFO_MAP_LATITUDE="37.405"
   MESH_INFO_MAP_LONGITUDE="-98.525"
   MESH_INFO_MAP_ZOOM="5"
   MESH_INFO_LOG_LEVEL=INFO
   MESH_INFO_MAP_TILE_URL="https://tiles.stadiamaps.com/tiles/stamen_terrain/{z}/{x}/{y}{r}.png"
   MESH_INFO_MAP_TILE_ATTRIBUTION='&copy; <a href="https://stadiamaps.com/" target="_blank">Stadia Maps</a> <a href="https://stamen.com/" target="_blank">&copy; Stamen Design</a> &copy; <a href="https://openmaptiles.org/" target="_blank">OpenMapTiles</a> &copy; <a href="https://www.openstreetmap.org/copyright" target="_blank">OpenStreetMap</a>'


Options
-------

.. note::

   The options for ``meshinfo report`` are configured on the command line.

MESH_INFO_COLLECTOR_LINK_INACTIVE
   Number of days a link remains on the map in inactive state after it was last seen.
   Default is 1 day.

MESH_INFO_COLLECTOR_NODE_INACTIVE
   Number of days a node remains on the map in inactive state after it was last seen.
   Default is 7 days.

MESH_INFO_COLLECTOR_TIMEOUT
   Total number of seconds to fetch ``sysinfo.json`` information before timing out.
   Default is 30.

MESH_INFO_COLLECTOR_WORKERS
   Number of worker tasks polling the network simultaneously.
   Default is 50.

MESH_INFO_DB_URL
   Change location of the SQLite database.
   Default is SQLite database ``mesh-info.db`` in the data directory.

MESH_INFO_ENVIRONMENT
   Either ``production`` or ``development``.
   Defaults to ``production``, use ``development`` to enable more debugging options.
   See :doc:`contributing` for more information.

MESH_INFO_LOCAL_NODE
   Specify the node name or IP address to connect to for OLSR information.
   Useful if *Mesh Info* server cannot resolve default of `localnode.local.mesh`.

MESH_INFO_LOG_LEVEL
   Controls how much information is logged/displayed in terminal.
   Can be one of the following (from most verbose to least):
   ``DEBUG``, ``INFO``, ``WARNING``, ``ERROR``.
   Default is ``WARNING``.

MESH_INFO_MAP_LATITUDE
   Latitude coordinate to center map on.

MESH_INFO_MAP_LONGITUDE
   Longitude coordinate to center map on.

MESH_INFO_MAP_MAX_ZOOM
   Set the max zoom level.
   Passed to Leaflet's `tileLayer()`_.
   Default is 18.

MESH_INFO_MAP_TILE_ATTRIBUTION
   The attribution to display on the map.
   Passed to Leaflet's `tileLayer()`_.

   .. warning::

      Because the attribution often contains HTML,
      special characters will *not* be escaped,
      and thus vulnerabilities could be introduced with a bad attribution.

MESH_INFO_MAP_TILE_URL
   The template URL for the tile server.
   Passed to Leaflet's `tileLayer()`_.

MESH_INFO_MAP_ZOOM
   Default zoom level.
   Passed to `Leaflet <https://leafletjs.com/>`_.

MESH_INFO_WEB_BIND
   TCP or Unix socket to bind web application to.
   Defaults to ``0.0.0.0:8000``.
   A Unix socket looks like ``unix:/run/mesh-info.sock``.


.. _tileLayer(): https://leafletjs.com/reference.html#tilelayer


CHANGELOG.rst
Changelog
=========

Changes to **Mesh Info** that affect users or are of major impact to developers.


The format is based on `Keep a Changelog <https://keepachangelog.com/en/1.0.0/>`_,
and this project adheres to `Semantic Versioning <https://semver.org/spec/v2.0.0.html>`_.

**NOTE:** This project is currently in development (pre-1.0),
so breaking changes are possible but will be highlighted here.

..
    Recommended Sections:

    Added
    Changed
    Deprecated
    Removed
    Fixed
    Security

0.8.0 - Unreleased
------------------

Changed
^^^^^^^

* Use updated RRDtool Python bindings package
* Switched development tooling to use PDM and Ruff

Removed
^^^^^^^

* **BREAKING CHANGE:** Removed support for PostgreSQL.

0.7.0 - 2024-05-20
------------------

Changed
^^^^^^^

* Bump pinned versions of several packages due to security vulnerabilities.
* Updated documentation and defaults for map tiles due to Stamen Tiles no longer being available (`#119 <https://github.com/smsearcy/mesh-info/pull/119>`_)

Fixed
^^^^^

* Fix issue with collector service hanging and not updating (`#124 <https://github.com/smsearcy/mesh-info/pull/124>`_)
* Updates for new version of AREDN firmware (`#123 <https://github.com/smsearcy/mesh-info/pull/123>`_)

0.6.0 - 2023-04-13
------------------

Added
^^^^^

* Add ``purge`` command to delete old data. (`#30 <https://github.com/smsearcy/mesh-info/issues/30>`_)
* Toggle to hide the link legend (`#84 <https://github.com/smsearcy/mesh-info/issues/84>`_)
* Support Python 3.11.
* Add iperf3 URL builder.  (`#72 <https://github.com/smsearcy/mesh-info/issues/72>`_)
* Use different Leaflet layers for different types of nodes/links.
* Links tab on the node popup on the map (`#70 <https://github.com/smsearcy/mesh-info/issues/70>`_).
* Tabs on link popup for cost and quality graphs (`#109 <https://github.com/smsearcy/mesh-info/pull/109>`_)

Changed
^^^^^^^

* **BREAKING CHANGE:** Dropped support for Python 3.7 and 3.8.
* Update RF link colors and style of infinite links (`#73 <https://github.com/smsearcy/mesh-info/issues/73>`_).
* **BREAKING CHANGE:** Logging levels ``TRACE`` and ``SUCCESS`` have been removed, ``WARNING`` is the new default (`#107 <https://github.com/smsearcy/mesh-info/issues/107>`_).
* **BREAKING CHANGE:** Renamed configuration variables for count of network pollers and polling timeout. (`#113 <https://github.com/smsearcy/mesh-info/issues/113>`_).

  * ``MESH_INFO_COLLECTOR_WORKERS`` replaces ``MESH_INFO_POLLER_MAX_CONNECTIONS``.
  * ``MESH_INFO_COLLECTOR_TIMEOUT`` replaces ``MESH_INFO_POLLER_CONNECT_TIMEOUT`` and ``MESH_INFO_POLLER_READ_TIMEOUT`` with a single, total timeout.

Fixed
^^^^^

* Bearings are always positive numbers (0 - 359).

0.5.1 - 2022-10-07
------------------

Fixed
^^^^^

* Corrected version typo.


0.5.0 - 2022-10-07
------------------

Added
^^^^^

* Add/expand project documentation. (`#28 <https://github.com/smsearcy/mesh-info/issues/28>`_)
* Basic implementation of network map.
  Client is required internet access to fetch tiles.
  All map elements rendered on single layer.
* Map legend (`#35 <https://github.com/smsearcy/mesh-info/issues/35>`_)
* Set starting map coordinates via configuration file. (`#45 <https://github.com/smsearcy/mesh-info/issues/45>`_)
* Save node polling errors to database. (`#55 <https://github.com/smsearcy/mesh-info/issues/55>`_)
* Display node polling errors. (`#47 <https://github.com/smsearcy/mesh-info/issues/47>`_)
* Configure map tile URL and attribution via configuration file.  (`#63 <https://github.com/smsearcy/mesh-info/issues/63>`_)

Changed
^^^^^^^

* Bumped current AREDN version (`#38 <https://github.com/smsearcy/mesh-info/issues/38>`_)
* Group nightly firmware versions and add API version statistics (`#42 <https://github.com/smsearcy/mesh-info/issues/42>`_)
* Cap link cost from API at 99.99 for consistency (`#81 <https://github.com/smsearcy/mesh-info/issues/81>`_)

Fixed
^^^^^

* Added new 2GHz channels (`#40 <https://github.com/smsearcy/mesh-info/issues/40>`_)
* Added new 5GHz channel (`#44 <https://github.com/smsearcy/mesh-info/issues/44>`_)
* Fix 3GHz icon color (`#49 <https://github.com/smsearcy/mesh-info/issues/49>`_)
* Fix link destination name issue (`#52 <https://github.com/smsearcy/mesh-info/issues/52>`_)
* Node list firmware sorted "naturally" (`#48 <https://github.com/smsearcy/mesh-info/issues/48>`_)
* Use node name for link to AREDN page (`#65 <https://github.com/smsearcy/mesh-info/issues/65>`_)


0.4.0 - 2022-06-02
------------------

Added
^^^^^

* Added SSID to the node table for searching/sorting (`#25 <https://github.com/smsearcy/mesh-info/issues/25>`_)

Changed
^^^^^^^

* **BREAKING CHANGE:** Renamed project and data folders (`#26 <https://github.com/smsearcy/mesh-info/issues/26>`_)

  * Renamed project from **pyMeshMap** to **Mesh Info**.
  * Renamed the Python package from ``pymeshmap`` to ``meshinfo``.
  * Data folders from ``pymeshmap` to ``mesh-info`` (in ``/var/lib/`` or ``~/.local/share/``).
  * The default SQLite database from ``pymeshmap.db`` to ``mesh-info.db``.
  * The GitHub repository from ``smsearcy/pymeshmap`` to ``smsearcy/mesh-info``.

* **BREAKING CHANGE:** Change default port for web service to 8000 (`#29 <https://github.com/smsearcy/mesh-info/issues/29>`_)

  Changed default port to 8000 (Gunicorn default)
  and configuration to use "bind" instead of "host" and "port"
  This enables binding to a Unix socket instead of a TCP port.

Fixed
^^^^^

* Fix import/export going into the ``rrd`` subdirectory with the RRD folders (`#19 <https://github.com/smsearcy/mesh-info/issues/19>`_)


0.3.0 - 2022-04-20
------------------

Added
^^^^^

* Start a changelog (`#21 <https://github.com/smsearcy/mesh-info/issues/21>`_)

Changed
^^^^^^^

* **BREAKING CHANGE:** Moved default data folder (`#18 <https://github.com/smsearcy/mesh-info/issues/18>`_)

  For *production*, moved from ``/usr/local/share/pymeshmap`` to ``/var/lib/pymeshmap``,

meshinfo/config.py
"""Configuration for Mesh Info from the environment."""

from __future__ import annotations

import enum
import logging
import os
from pathlib import Path
from typing import Any, Dict, Optional

import attrs
import environ
import pendulum
import platformdirs
import structlog
from dotenv import load_dotenv
from pyramid.config import Configurator

from .aredn import VersionChecker
from .historical import HistoricalStats

logger = structlog.get_logger()

FOLDER_NAME = "mesh-info"


class Environment(enum.Enum):
    DEV = "development"
    PROD = "production"


def default_workers():
    """Defaults to Gunicorn recommendations."""
    if cpu_count := os.cpu_count():
        return cpu_count * 2 + 1
    else:
        return 1


def _get_log_level(level: str) -> int:
    return getattr(logging, level)


@environ.config(prefix="MESH_INFO")
class AppConfig:
    @environ.config
    class Aredn:
        current_firmware: str = environ.var(default="3.24.4.0")
        current_api: str = environ.var(default="1.13")

    @environ.config
    class Collector:
        workers: int = environ.var(default=50, converter=int)
        timeout: int = environ.var(default=30, converter=int)
        node_inactive: int = environ.var(default=7, converter=int)
        link_inactive: int = environ.var(default=1, converter=int)
        period: int = environ.var(default=5, converter=int)

    @environ.config
    class DB:
        url: str = environ.var(default="")
        pool_pre_ping: bool = environ.bool_var(default=True)

    @environ.config
    class Map:
        latitude: float = environ.var(default=19.64, converter=float)
        longitude: float = environ.var(default=-1.58, converter=float)
        zoom: int = environ.var(default=3, converter=int)
        max_zoom: int = environ.var(default=18, converter=int)
        tile_url: str = environ.var(default="")
        tile_attribution: str = environ.var(default="")

    @environ.config
    class Web:
        bind: str = environ.var(default="0.0.0.0:8000")
        workers: int = environ.var(default=default_workers(), converter=int)
        # debug_hosts

    env: Environment = environ.var(default="production", converter=Environment)
    local_node: str = environ.var(default="localnode.local.mesh")
    log_level: int = environ.var(default="WARNING", converter=_get_log_level)
    site_name: str = environ.var(default="Mesh Info")
    data_dir: Path = environ.var(default="")

    aredn: Aredn = environ.group(Aredn)
    collector: Collector = environ.group(Collector)
    db: DB = environ.group(DB)
    map: Map = environ.group(Map)
    web: Web = environ.group(Web)

    def __attrs_post_init__(self):
        if self.data_dir == "":
            # default data directory depends on environment
            if self.env == Environment.PROD:
                self.data_dir = Path(f"/var/lib/{FOLDER_NAME}")
            elif self.env == Environment.DEV:
                self.data_dir = platformdirs.user_data_path(FOLDER_NAME) / "data"
        else:
            self.data_dir = Path(self.data_dir)

        if self.db.url == "":
            # location of default SQLite database depends on the data_dir
            self.db.url = f"sqlite:///{self.data_dir!s}/mesh-info.db"

        if self.env == Environment.DEV:
            # Only use 1 worker in development environment for the debug toolbar
            self.web.workers = 1

    @property
    def rrd_dir(self) -> Path:
        """Directory for RRD files under the data directory."""
        return self.data_dir / "rrd"


def from_env() -> AppConfig:
    # walks up the folder path looking for `.env` file
    # and loads into environment variables
    # (which we then put into the settings dictionary)
    load_dotenv()
    return environ.to_config(AppConfig)


def configure(
    settings: Optional[Dict[str, Any]] = None, *, app_config: Optional[AppConfig] = None
) -> Configurator:
    """Configure the Pyramid application."""

    if settings is None:
        settings = {}

    if app_config is None:
        app_config = from_env()

    # This is kind of ugly since I'm mashing together
    # (I should update my custom stuff to just use the AppConfig object)
    settings["app_config"] = app_config

    # Configure settings for Pyramid
    settings["pyramid.retry"] = 3

    # Configure Jinja2
    settings["jinja2.directories"] = "meshinfo:templates"
    settings["jinja2.filters"] = {
        "duration": "meshinfo.filters.duration",
        "in_tz": "meshinfo.filters.in_tz",
        "local_tz": "meshinfo.filters.local_tz",
        "unknown": "meshinfo.filters.unknown",
    }

    if app_config.env == Environment.DEV:
        settings["debugtoolbar.max_visible_requests"] = 25
        settings["pyramid.debug_authorization"] = False
        settings["pyramid.debug_notfound"] = False
        settings["pyramid.debug_routematch"] = False
        settings["pyramid.prevent_cachebust"] = True
        settings["pyramid.reload_all"] = True

    # configure logging
    configure_logging(app_config.log_level)

    logger.debug(
        "Application configuration", **attrs.asdict(app_config)  # type: ignore
    )

    # configure Pyramid application
    config = Configurator(settings=settings)

    config.include("pyramid_retry")
    config.include("pyramid_services")
    config.include("pyramid_jinja2")
    config.include(".routes")
    config.include(".models")

    if app_config.env == Environment.DEV:
        config.include("pyramid_debugtoolbar")

    server_timezone = pendulum.tz.local_timezone()

    def client_timezone(request):
        if "local_tz" in request.cookies:
            try:
                client_tz = pendulum.timezone(request.cookies["local_tz"])
            except Exception as exc:
                logger.warning(
                    "Invalid timezone specified",
                    local_tz=request.cookies["local_tz"],
                    error=repr(exc),
                )
                client_tz = server_timezone
            return client_tz.name
        else:
            return server_timezone.name

    config.add_request_method(lambda r: client_timezone(r), "timezone", reify=True)

    # Register services with `pyramid-services`

    # Register the `VersionChecker` singleton
    version_checker = VersionChecker.from_config(app_config.aredn)
    config.register_service(version_checker, VersionChecker)

    # Register the `HistoricalStats` singleton
    config.register_service(
        HistoricalStats(data_path=app_config.rrd_dir), HistoricalStats
    )

    config.scan(".views")

    config.commit()
    return config


def configure_logging(level: int):
    """Configure structlog with the specified level."""
    processors: list = [
        structlog.contextvars.merge_contextvars,
        structlog.processors.add_log_level,
        structlog.processors.StackInfoRenderer(),
        structlog.dev.set_exc_info,
    ]
    if "INVOCATION_ID" not in os.environ:
        # add timestamps when not running in systemd
        processors.append(structlog.processors.TimeStamper(fmt="iso", utc=False))
    # TODO: do we need a way to disable colors?  (doesn't work with redirecting output)
    processors.append(structlog.dev.ConsoleRenderer())

    structlog.configure(
        processors=processors,
        wrapper_class=structlog.make_filtering_bound_logger(level),
        context_class=dict,
        logger_factory=structlog.PrintLoggerFactory(),
        cache_logger_on_first_use=False,
    )
